File: services/message_broker/internal/envelope/envelope.go
package envelope

import (
	"bytes"
	"encoding/json"
	"fmt"
)

// EngineMessage mirrors governance/schemas/engine-contract-v1.json.
type EngineMessage struct {
	ContractVersion string          `json:"contract_version"`
	MessageType     string          `json:"message_type"`
	SourceEngine    string          `json:"source_engine"`
	Timestamp       string          `json:"timestamp"`
	Payload         json.RawMessage `json:"payload"`
	Metadata        MessageMetadata `json:"metadata"`
}

type MessageMetadata struct {
	SpecificationID  string `json:"specification_id"`
	PolicyVersion    string `json:"policy_version"`
	Owner            string `json:"owner"`
	ValidationStatus string `json:"validation_status"`
}

var validEngines = map[string]bool{
	"rust_engine": true, "python_engine": true, "go_engine": true,
}

func Validate(data []byte) (*EngineMessage, error) {
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.DisallowUnknownFields()
	var msg EngineMessage
	if err := dec.Decode(&msg); err != nil {
		return nil, fmt.Errorf("envelope decode: %w", err)
	}
	if msg.ContractVersion != "1.0" {
		return nil, fmt.Errorf("invalid contract_version: %s", msg.ContractVersion)
	}
	if !validEngines[msg.SourceEngine] {
		return nil, fmt.Errorf("invalid source_engine: %s", msg.SourceEngine)
	}
	if msg.MessageType == "" || msg.Timestamp == "" {
		return nil, fmt.Errorf("missing required fields")
	}
	if msg.Metadata.SpecificationID == "" || msg.Metadata.Owner == "" {
		return nil, fmt.Errorf("missing required metadata fields")
	}
	return &msg, nil
}

File: core_engine/risk/src/envelope.rs
use serde::{Deserialize, Serialize};
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct EngineMessage<T: Serialize> {
    pub contract_version: String,
    pub message_type: String,
    pub source_engine: String,
    pub timestamp: String,
    pub payload: T,
    pub metadata: MessageMetadata,
}
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MessageMetadata {
    pub specification_id: String,
    pub policy_version: String,
    pub owner: String,
    pub validation_status: String,
}
impl<T: Serialize> EngineMessage<T> {
    pub fn new_risk_message(
        message_type: impl Into<String>,
        specification_id: impl Into<String>,
        payload: T,
    ) -> Self {
        Self {
            contract_version: "1.0".to_string(),
            message_type: message_type.into(),
            source_engine: "rust_engine".to_string(),
            timestamp: chrono::Utc::now().to_rfc3339(),
            payload,
            metadata: MessageMetadata {
                specification_id: specification_id.into(),
                policy_version: "1.0".to_string(),
                owner: "Risk Engine".to_string(),
                validation_status: "validated".to_string(),
            },
        }
    }
}
